src/Microsoft.Extensions.ML/ModelReloadToken.cs

```diff
@@ -44,6 +44,12 @@ public ModelReloadToken()
         /// <summary>
         /// Used to trigger the change token when a reload occurs.
         /// </summary>
-        public void OnReload() => _cts.Cancel();
+        public void OnReload()
+        {
+            if (!_cts.IsCancellationRequested)
+                _cts.Cancel();
```

Comment:
The code seems to be saying if cancellation is not requested, call Cancel(). Am I reading that correctly? Can you please describe the root cause and the fix a bit more thoroughly in the pull request?

Reply:
This means if the cancellation has not already been requested, then cancel the token.
we should always add this check as calling cancel twice will crash the process. This is one of the reason that test process is crashing.

---
In reply to: [350532410](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350532410) [](ancestors = 350532410)

Reply:
> This is one of the reason that test process is crashing.

Can you provide a test log that shows this is where the test process is crashing? I don't see `.Cancel()` causing a process crash.

I made a console app that calls Cancel multiple times, and it doesn't even throw an exception.

Reply:
Below is the call stack I have reproduced in my local env:

The active test run was aborted. Reason: Test host process crashed : Unhandled Exception: System.AggregateException: One or more errors occurred. (Safe handle has been closed) ---> System.ObjectDisposedException: Safe handle has been closed
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Microsoft.Win32.Win32Native.SetEvent(SafeWaitHandle handle)
   at System.Threading.EventWaitHandle.Set()
   at Microsoft.Extensions.ML.FileLoaderTests.<>c__DisplayClass2_0.<can_reload_model>b__1() in C:\work\machinelearning\test\Microsoft.Extensions.ML.Tests\FileLoaderTests.cs:line 65
   at Microsoft.Extensions.Primitives.ChangeToken.<>c__DisplayClass0_0.<OnChange>b__0(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   --- End of inner exception stack trace ---
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   at Microsoft.Extensions.ML.ModelReloadToken.OnReload() in C:\work\machinelearning\src\Microsoft.Extensions.ML\ModelReloadToken.cs:line 47
   at Microsoft.Extensions.ML.FileModelLoader.WatcherChanged(Object sender, FileSystemEventArgs e) in C:\work\machinelearning\src\Microsoft.Extensions.ML\ModelLoaders\FileModelLoader.cs:line 78
   at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(WatcherChangeTypes changeType, String name)
   at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer)
   at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)


---
In reply to: [350906884](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350906884) [](ancestors = 350906884)

Reply:
> we should always add this check as calling cancel twice will crash the process. This is one of the reason that test process is crashing.

This is not correct. Cancellation a second time is a NOP, and `Cancel` itself is guarded by a check of `IsCancellationRequested`.